### ISSUE TITLE
problem with uncertainty calculation

### ISSUE THREAD

**Issue by TMullerSG** [kumar-shridhar/PyTorch-Softplus-Normalization-Uncertainty-Estimation-Bayesian-CNN#2]
def calc_uncertainty_normalized(output):
    outputs = []
    for t in range(1):
        prediction = F.softplus(output.cpu())
        prediction = normalization_function(prediction)
        outputs.append(prediction)
        
    res = np.mean(prediction.numpy(), axis=0)
    p_hat= torch.cat(outputs, 1)
    p_hat=p_hat.numpy()
    T=1
    
    aleatoric = np.diag(res) - p_hat.T.dot(p_hat)/p_hat.shape[0]
#     aleatoric += aleatoric
    tmp = p_hat - res  
    epistemic = tmp.T.dot(tmp)/tmp.shape[0]
#     epistemic += epistemic 

    print(np.sum(epistemic, keepdims = True))
    print(np.sum(aleatoric, keepdims = True))
    
    
    
    

#     p_hat = np.array(res)
#     print (p_hat)
#     epistemic = np.mean((p_hat) ** 2, axis=0) - np.mean((p_hat), axis=0) ** 2
#     epistemic = np.mean((1-p_hat) ** 2, axis=0) - np.mean((1-p_hat), axis=0) ** 2
#    epistemic += epistemic 
#     #print (epistemic)
#     aleatoric = np.mean((p_hat) * (1-(p_hat)), axis = 0)
#     aleatoric = np.mean((1-p_hat), axis = 0) ** 2
#     aleatoric += aleatoric
    #print (aleatoric)
    return (np.sum(epistemic, keepdims = True)), (np.sum(epistemic, keepdims = True))

is this the final code you use? the last return terms are two epistemic

**Comment by Jordy-VL** [comment]
Yeah I am similarly in doubt about the correctness of the code, and the formulas mentioned in the paper.

**Comment by romainVala** [comment]
Hi
did you figure out some correction ? (or found other implementation ?)
Many thanks

**Comment by Jordy-VL** [comment]
I implemented the methods from https://arxiv.org/pdf/1811.07253
<https://arxiv.org/pdf/1811.07253.pdf>.
These are fairly similar to the ones presented by Kendall & Gal (2017).
I will soon (middle of June) release my code jointly with a research paper,
so keep on the watch for posts on my Github repos.
If you have any specific questions, then we can look at it together maybe?
